### DISCUSSION TITLE
API Functions & The Need for Lower Level Abstractions

### DISCUSSION THREAD

**Discussion by yaq-cc** [pdfcpu/pdfcpu#877] [General]
Hi pdfcpu maintainers,

Great job with the tool and the library; I'm very impressed how much thought has gone into this.  I have a use case where I need to split PDFs and write to some form of cloud based object storage.

I reviewed pkg.go.dev and I found some functions (the Split family of functions - https://pkg.go.dev/github.com/pdfcpu/pdfcpu/pkg/api#Split) that would make this easy.  These functions are wired up to accept a local directory to write those files to which means I need to write to local disk, find those files, and then perform something like io.Copy to each one of my io.Writers for the split files, which feels unnecessary.

Is there any way to get a slice of io.Readers or an iterator of io.Readers from the splits instead so I can stream to the backends of my choice?  I did some (*light*) diving into the source code and a lot of the methods are unexported.  

Additionally: 

- what is the "naming" or "renaming" policy of the PDF splits?
- One of the libraries I use in Python is called PyPDF3.  This library gives the caller the ability to implement Deflate compression and remove links on the PDF to shrink the file size - does pdfcpu have similar methods / tools?  please let me know!

Regards,

**Comment by hhrutter** [comment]
Hello!

Decoupling the processing step from consuming any command output is work in progress.
We need this consistently across the board in order to go Beta.
I am sorry this may not be obvious but it is in the pipeline.

As far as naming policies right now please take a look at `pdfcpu help split` on the CLI.

As well check out `pdfcpu help annotation` in order to remove eg. all link annotations in a file.

`Deflate` is a standard compression technique in PDF.
What do you mean by `implementing Deflate compression` ?
Please elaborate.

If we deep dive into one of your issues please open separate dedicated threads or you can always create an issue.

Please consider becoming a pdfcpu sponsor.
Sponsorship ensures the very future and continued progress of the project and comes with prioritized individual support.

Thanks for considering pdfcpu 💚

**Comment by hhrutter** [comment]
Closing this as resolved.

For splitting without writing intermediate files to disk, the Go API now exposes api.SplitRaw, which returns page spans carrying io.Readers:

ps, err := api.SplitRaw(rs, span, conf)

Each returned PageSpan contains From, Thru, and Reader, so callers can stream the generated split PDFs directly to object storage or any other backend of their choice.

For file-based workflows, Split / SplitFile remain available. 
For backend integrations, SplitRaw is the lower-level abstraction this discussion was asking for.
